WordPress/openverse-api
# Cache waveform data in database

## Fixes


Fixes #490 by @sarayourfriend

## Description


This PR indefinitely caches waveform data in the postgres DB directly on the waveform model after the first time it is generated.

## Testing Instructions

I added a thorough unit test to test the `get_or_create_waveform` function. There were no existing integration tests for the waveform endpoint. Should we add some? I'd appreciate some guidance about how best to do that and prevent the call to the `audiowaveform` binary and an external call to retrieve the audio file (can we mock the return of the audio file from `requests.get`?)

You can test this manually by running the API locally following the instructions in the README and then visiting the http://localhost:8000/v1/audio/010356c6-ff57-44f1-bddc-675a2a1e5a38/waveform/ endpoint. The first time should take noticably longer than subsequent visits.

## Checklist


[best_practices]:https://git-scm.com/book/en/v2/Distributed-Git-Contributing-to-a-Project#_commit_guidelines

## Developer Certificate of Origin


<details>
<summary>Developer Certificate of Origin</summary>

```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
```

</details>
